### ISSUE TITLE
Line chart. Event annotation not work

### ISSUE THREAD

**Issue by slavanorm** [apache/superset#24374]
Line chart. Event annotation not work.
error raised from python (raises every time):
![telegram-cloud-photo-size-4-5769422933448572157-y](https://github.com/apache/superset/assets/48209342/2be25cd6-fdeb-4c86-b655-547458d220b3)

#### How to reproduce the bug

1. create a line chart
2. add Event annotation
3. click apply/ok and get the error

#### Screenshots
<img width="1379" alt="Screenshot 2023-06-13 at 10 17 06" src="https://github.com/apache/superset/assets/48209342/60992ffc-0f38-4949-a83d-402af4871b7a">

### Environment

(please complete the following information):

- browser type and version: chrome v 114.0.5375
- superset version: `2.1.0`
- python version: `3.10`

### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [v] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [v] I have reproduced the issue with at least the latest released version of superset.
- [v] I have checked the issue tracker for the same issue and I haven't found one similar.

### Additional context
SQL server: psql 14
failed annotation with "superset annotation" and "table", same error.

works on 1.5.3, so feels like recent changes are breaking

**Comment by slavanorm** [comment]
idk.. i created p1d and P1D cols and it still not working

**Comment by Mavyre** [comment]
Same here but with 'P1M' column. Event Annotation on Mixed Chart.
I can't understand what's going on, and sometimes it start to magically work

**Comment by azamrik** [comment]
This issue seems to happen when adding annotations to a chart when:

- It has time grain (p1d is daily grain, p1w is weekly, p1m is monthly, ...etc)
- The chart is already part of a dashboard


A work around that worked for me is to recreate the chart with the annotation and then add it to the dashboard.

**Comment by Mavyre** [comment]
> This issue seems to happen when adding annotations to a chart when:
> 
> * It has time grain (p1d is daily grain, p1w is weekly, p1m is monthly, ...etc)
> * The chart is already part of a dashboard
> 
> A work around that worked for me is to recreate the chart with the annotation and then add it to the dashboard.

Thanks for the comment. Gave me an idea that actually works: Checking "Override Time Grain" does the job for me, even if the chart already is part of a dashboard!

**Comment by aagorelik** [comment]
The issue still persists (helm version `0.11.2`), even when using Superset own Annotations Layer and Time Grain removed from the chart:

![image](https://github.com/apache/superset/assets/1843995/fc0aad12-0a4f-4e33-8fc2-a17a5e0fe0d0)

Or when used own table (with, or without `P1D` field added; and with, or without Time Grain):

![image](https://github.com/apache/superset/assets/1843995/261c165b-6d34-42c5-b58e-3ec7aba9d7c1)

![image](https://github.com/apache/superset/assets/1843995/beac013a-781f-4548-982c-ee4f0af1d7e8)

**Comment by oliverlambson** [comment]
Same issue here 🤚

**Comment by konvica** [comment]
Same issue. What solved it for me, was to recreate the annotations table outside of any dashboard and set time grain in the annotation table to "minute".  For the line chart I use time grain "hour" or "minute".

**Comment by rusackas** [comment]
Seems we have a couple workarounds, and I'm wondering if this is enough to close the issue. Since it's been silent for upward of 6 months, it's at risk of being closed as stale. Note that soon we'll also only support Superset 4.0 and up, once 4.1 is out. If people can make sure this is still an issue (it probably is) in 4.0, that'd be appreciated.

**Comment by sfirke** [comment]
Using 4.0.2, I just heard from a user today who has a problem with annotation layers sometimes adding, sometimes not, to the same chart.  I was able to replicate it though I don't understand when it fails and when it succeeds.  We don't get any error message at all, maybe that's because we're using a saved Superset annotation source instead of a table?  We just can't click "apply" or "OK" on the modal.  Seems related to this issue, maybe has the same underlying problem.

Just now I was unable to add it to a chart.  I refreshed my browser and then I could add it 😕

**Comment by rusackas** [comment]
I'm assuming this is still just as much of an issue today, but holler if anyone's perceived any change.

**Comment by rusackas** [comment]
Agreed the annotation feature needs work (or an entire revamp). Any volunteers that want to fix this issue are welcome to do so, but otherwise this might be closed as "not planned" until someone (anyone) opens a SIP to propose a revised feature design and implementation.

**Comment by Arunodoy18** [comment]
Thanks for all the detailed reports and workarounds shared here — very helpful context.

I’m interested in investigating this further to better understand the root cause, especially around how time grain (e.g. P1D/P1M) interacts with annotation layers and chart queries.

Before attempting any fix, I plan to:
- Reproduce this on a recent Superset version
- Trace how time grain fields are resolved between chart query generation and annotation layer queries
- Check whether this is a query generation issue, datasource metadata issue, or frontend state inconsistency

If maintainers have any pointers on where the most likely root cause lives (annotation query building, time grain column mapping, or datasource schema handling), that would be really helpful.

Happy to share findings first before proposing any code change.

**Comment by rusackas** [comment]
Circling back on this one. The annotation failure here is pretty diffuse (non-deterministic across versions/DBs, and tangled up in how older saved charts store the time grain), so it's not really a clean point-fix. As I mentioned earlier, I think this needs a SIP to redesign the annotation flow rather than another band-aid (the earlier attempt in #37593 went the regex route and stalled). Going to close as not-planned for now, but happy to reopen if someone picks up a SIP. Thanks all for the reports and patience here.
